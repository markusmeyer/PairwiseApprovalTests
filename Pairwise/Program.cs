using System;
using System.Collections.Generic;
using System.Text;
using ApprovalTests;
using ApprovalTests.Combinations;
using ApprovalTests.Reporters;
using NUnit.Framework;
using NUnit.Framework.Internal.Builders;

namespace Pairwise
{
	[TestFixture]
	[UseReporter(typeof(DiffReporter))]
	class Program
	{
		[Test]
		public void CombinationTest()
		{
			CombinationApprovals.VerifyAllCombinations(
				ParameterizedTest,
				new[]{ 1, 2, 3, 4 },
				new[]{ 2, 3, 4 },
				new[]{ 3, 4, 5 });
		}

		[Test]
		public void PairwiseTest()
		{
			PairwiseApprovals.VerifyAllPairs(
				ParameterizedTest,
				new[]{ 1, 2, 3, 4 },
				new[]{ 2, 3, 4 },
				new[]{ 3, 4, 5 });
		}

		object ParameterizedTest(int a, int b, int c)
		{
			return new ClassUnderTest().MethodUnderTest(a, b, c);
		}
	}

	static class PairwiseApprovals
	{
		public static void VerifyAllPairs<A,B,C>(
			Func<A,B,C,object> function,
			IEnumerable<A> aItems,
			IEnumerable<B> bItems,
			IEnumerable<C> cItems)
		{
			var testCases = new PairwiseStrategy().GetTestCases(new System.Collections.IEnumerable[]
			{
				aItems,
				bItems,
				cItems
			});
			StringBuilder output = new StringBuilder();
			foreach (NUnit.Framework.Interfaces.ITestCaseData testCase in testCases)
			{
				var result = function(
					(A)testCase.Arguments[0],
					(B)testCase.Arguments[1],
					(C)testCase.Arguments[2]);
				output.AppendLine(
					"[" + string.Join(",", testCase.Arguments) + "] => " +
					result);
			}
			Approvals.Verify(output);
		}
	}

	class ClassUnderTest
	{
		public int MethodUnderTest(int a, int b, int c)
		{
			return a * b + c;
		}
	}
}