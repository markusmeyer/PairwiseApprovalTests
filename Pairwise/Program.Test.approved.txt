1, 4, 3
1, 2, 5
1, 3, 4
2, 3, 5
2, 2, 3
2, 4, 4
3, 2, 4
3, 3, 3
3, 4, 5
4, 3, 5
4, 2, 4
4, 4, 3
